### ISSUE TITLE
Forecasting with extra regressors

### ISSUE THREAD

**Issue by crhistian123** [facebook/prophet#1455]
Hi everyone. First of all, I want to make clear that I'm new at R language and even newbier in prophet. 

I am making some vehicles volumes with the Prophet package at Rstudio and already had great results for daily volumes in a specified highway in my country. But, after this forecast, I want to include extra regressors such as temperature, rain, days between two holidays, etc. However, I've been having some dificult. Here's my code:

library(prophet)
library(dplyr)
library(xlsx)
library(utils)

```
#Vehicle Volume
names(data)[2]="y"
Vol_Passeio=data[,2]
Vol_Passeio

#Dates
Df=data.frame(ds=seq(as.Date('2010-01-01'),as.Date('2018-12-31'),by='d'),
              y=Vol_Passeio)
Df

#Holidays
#----------------------------------------------------------------------------------------------------------------

#Feriado 1º de janeiro
Ano_Novo=data.frame(
  holiday='Ano novo',
  ds=as.Date(c('2010-01-01','2011-01-01','2012-01-01','2013-01-01','2014-01-01','2015-01-01','2016-01-01','2017-01-01','2018-01-01','2019-01-01','2020-01-01','2021-01-01','2022-01-01')),
  lower_window=-6,
  upper_window=0
)
Ano_Novo

#Feriado de carnaval
Carnaval_BR=  data.frame(
  holiday='Carnaval',
  ds=as.Date(c('2010-02-16','2011-03-08','2012-02-21','2013-02-12','2014-03-04','2015-02-17','2016-02-09','2017-02-28','2018-02-13','2019-03-05','2020-02-25','2021-02-16','2022-03-01')),
  lower_window=-5,
  upper_window=1
)
Carnaval_BR

#Anversário de São Paulo
Aniversario_SP=data.frame(
  holiday='Aniversário de SP',
  ds=as.Date(c('2010-01-25','2011-01-25','2012-01-25','2013-01-25','2014-01-25','2015-01-25','2016-01-25','2017-01-25','2018-01-25','2019-01-25','2020-01-25','2021-01-25','2022-01-25')),
  lower_window=-1,
  upper_window=0
)

#Páscoa (Pela série histórica, há um aumento no volume entre a quarta feira anterior e o domingo de páscoa)
Pascoa=data.frame(
  holiday='Páscoa',
  ds=as.Date(c('2010-04-04','2011-04-24','2012-04-08','2013-03-31','2014-04-20','2015-04-05','2016-03-27','2017-04-16','2018-04-01','2019-04-21','2020-04-12','2021-04-04','2022-04-17')),
  lower_window=-4,
  upper_window=0
  )

#Tiradentes (A influcência varia de acordo com o dia da semana que cai o feriado. Provavelmente as dummies irão explicar esse comportamento)
Tiradentes=data.frame(
  holiday='Tiradentes',
  ds=as.Date(c('2010-04-21','2011-04-21','2012-04-21','2013-04-21','2014-04-21','2015-04-21','2016-04-21','2017-04-21','2018-04-21','2019-04-21','2020-04-21','2021-04-21','2022-04-21')),
  lower_window=-1,
  upper_window=0
)

#Dia do Trabalho
Trabalho=data.frame(
  holiday='Trabalho',
  ds=as.Date(c('2010-05-01','2011-05-01','2012-05-01','2013-05-01','2014-05-01','2015-05-01','2016-05-01','2017-05-01','2018-05-01','2019-05-01','2020-05-01','2021-05-01','2022-05-01')),
  lower_window=-1,
  upper_window=0
)

#Corpus Christi
Corpus_Christi=data.frame(
  holiday='Corpus_Christi',
  ds=as.Date(c('2010-06-03','2011-06-23','2012-06-07','2013-05-30','2014-06-19','2015-06-04','2016-05-26','2017-06-15','2018-05-31','2019-06-20','2020-06-11','2021-06-03','2022-04-16')),
  lower_window=-4,
  upper_window=0
  )

#Revolução
Revolucao=data.frame(
  holiday='Revolução',
  ds=as.Date(c('2010-07-09','2011-07-09','2012-07-09','2013-07-09','2014-07-09','2015-07-09','2016-07-09','2017-07-09','2018-07-09','2019-07-09','2020-07-09','2021-07-09','2022-07-09')),
  lower_window=-1,
  upper_window=0
)

#Independência do Brasil
Independencia=data.frame(
  holiday='Independêncaia',
  ds=as.Date(c('2010-09-07','2011-09-07','2012-09-07','2013-09-07','2014-09-07','2015-09-07','2016-09-07','2017-09-07','2018-09-07','2019-09-07','2020-09-07','2021-09-07','2022-09-07')),
  lower_window=-1,
  upper_window=0
)

#Nossa Senhora Aparecida
NSA=data.frame(
  holiday='NSA',
  ds=as.Date(c('2010-10-12','2011-10-12','2012-10-12','2013-10-12','2014-10-12','2015-10-12','2016-10-12','2017-10-12','2018-10-12','2019-10-12','2020-10-12','2021-10-12','2022-10-12')),
  lower_window=-1,
  upper_window=0
)

#Finados
Finados=data.frame(
  holiday='Finados',
  ds=as.Date(c('2010-11-02','2011-11-02','2012-11-02','2013-11-02','2014-11-02','2015-11-02','2016-11-02','2017-11-02','2018-11-02','2019-11-02','2020-11-02','2021-11-02','2022-11-02')),
  lower_window=-1,
  upper_window=0
)

#Proclamação da República
Proclamacao=data.frame(
  holiday='Proclamação da Rep.',
  ds=as.Date(c('2010-11-15','2011-11-15','2012-11-15','2013-11-15','2014-11-15','2015-11-15','2016-11-15','2017-11-15','2018-11-15','2019-11-15','2020-11-15','2021-11-15','2022-11-15')),
  lower_window=-1,
  upper_window=0
)

#Consciência Negra
Consciencia_Negra=data.frame(
  holiday='Consciência Negra',
  ds=as.Date(c('2010-11-20','2011-11-20','2012-11-20','2013-11-20','2014-11-20','2015-11-20','2016-11-20','2017-11-20','2018-11-20','2019-11-20','2020-11-20','2021-11-20','2022-11-20')),
  lower_window=-1,
  upper_window=0
)

#Natal
Natal=data.frame(
  holiday='Natal',
  ds=as.Date(c('2010-12-25','2011-12-25','2012-12-25','2013-12-25','2014-12-25','2015-12-25','2016-12-25','2017-12-25','2018-12-25','2019-12-25','2020-12-25','2021-12-25','2022-12-25')),
  lower_window=-3,
  upper_window=0
)

#Criando o data frame de feriados
holidays=bind_rows(Ano_Novo,Aniversario_SP,Carnaval_BR,Pascoa,Tiradentes,Trabalho,Corpus_Christi,Revolucao,Independencia,NSA,Finados,Proclamacao,Consciencia_Negra,Natal)
holidays

#PROPHET
#-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
m=prophet(Df,holidays = holidays)
m
m$train.holiday.names

prev=make_future_dataframe(m,periods=365)
prev

forecast=predict(m,prev)
forecast

prophet_plot_components(m, forecast)

#Gráfico
dyplot.prophet(m,forecast)

#Excel
write.xlsx(forecast,"14.6 - Natal - Ano novo - Influência -6 e +4.xlsx")



#Extra regressors
#Importing days of the week
Sunday=BD_Sunday[1:3287,]
names(Sunday)[2]="Dom"


#Adicionando variáveis regressoras à base de dados original
Df_com_regressores=add_regressor(Df,Domingo,name='Dom')
```

But after the add_regressor(), I had this error

> Error in if (!(mode %in% c("additive", "multiplicative"))) { : 
>   argument have length zero
> Além disso: Warning message:
> In if (prior.scale <= 0) { :
>   the condition have comprimento > 1 and only the first element will be used

How can I include the Sunday extra regressor in my data frame Df? Attached this question there are some pictures of the respective data frames.

Thanks!!

![Df](https://user-images.githubusercontent.com/61809735/80384871-ce329d80-887b-11ea-9c83-d8165a0034a4.png)
![sunday data frame](https://user-images.githubusercontent.com/61809735/80384874-cecb3400-887b-11ea-8e9d-8ccd1dbf9595.png)

**Comment by bletham** [comment]
You need to add the extra regressor before fitting, and the function takes in the model object first, not the dataframe. If you do `?add_regressor` it will tell you what the inputs and outputs of the function are. See here for how to use `add_regressor`: https://facebook.github.io/prophet/docs/seasonality,_holiday_effects,_and_regressors.html#additional-regressors
In your data, it would look something like
```
m <- prophet()
m <- add_regressor(m, 'Dom')
m <- fit.prophet(m, Df, holidays = holidays)
```

I will note though that adding Sunday as an extra regressor should be unnecessary; the weekly seasonality will already be fitting a Sunday effect, so this would be redundant to that.

**Comment by crhistian123** [comment]
Hi, @bletham . Thank you for your time.

Late in that day I accomplished the forecasting with extra regressors and yes, the regressors related with days of the week are not necessary and didn't contributed to my model.
However, I include another extra regressor like days between holiday and weekens just like you said and it works. Thank you very much.
I will describe below how I did just for the future, in case that someone have the same doubt:

```
m <- prophet()
m<-prophet(holidays = holidays)
m <- add_regressor(m, 'Dom')
m<- add_regressor(m, 'extra_regressor_2')
m<- add_regressor(m, 'extra_regressor_3')
m <- fit.prophet(m, Df)
```

Just want to mention that is necessary to include in the original dataframe (Df) and in the prev dataframe (the one wth the forecasted dates) the extra regressors as a columns with the same name in both cases.

Thanks!
